src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1374,41 +1390,87 @@ RouterJob(Query *originalQuery, RelationRestrictionContext *restrictionContext,
 		return job;
 	}
 
-	pg_get_query_def(originalQuery, queryString);
-
-	if (originalQuery->commandType == CMD_SELECT)
+	/*
+	 * If it's a multiple shard update/delete query we need to create task for
+	 * each shard placement.
+	 */
+	if (*multiShardModifyQuery)
 	{
-		task = CreateTask(ROUTER_TASK);
+		List *taskList = NIL;
+		ListCell *shardPlacementCell = NULL;
+		uint64 jobId = INVALID_JOB_ID;
+		int taskId = 1;
+
+		/* lock metadata before getting placement lists */
+		LockShardListMetadata(placementList, ShareLock);
+
+		foreach(shardPlacementCell, placementList)
+		{
+			ShardPlacement *shardPlacement = (ShardPlacement *) lfirst(
+				shardPlacementCell);
+			ShardInterval *shardInterval = LoadShardInterval(shardPlacement->shardId);
+			Oid relationId = shardInterval->relationId;
+			uint64 shardId = shardInterval->shardId;
+			StringInfo shardQueryString = makeStringInfo();
+			Task *task = NULL;
+
+			deparse_shard_query(originalQuery, relationId, shardId, shardQueryString);
+
+			task = CitusMakeNode(Task);
+			task->jobId = jobId;
+			task->taskId = taskId++;
+			task->taskType = MODIFY_TASK;
+			task->queryString = shardQueryString->data;
+			task->dependedTaskList = NULL;
+			task->replicationModel = REPLICATION_MODEL_INVALID;
+			task->anchorShardId = shardId;
+			task->taskPlacementList = FinalizedShardPlacementList(shardId);
```

Comment:
We generally shouldn't assign placements in the planner. We do this in `CitusModifyBeginScan`.